### PR TITLE
FM-252: Remove duplicated code from client and handler classes

### PR DESCRIPTION
**Pull request by oskarhinc**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java [hunk 1]
```diff
@@ -1,11 +1,234 @@
 package org.openmrs.module.fhir.api.diagnosticreport.handler;
 
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.hl7.fhir.dstu3.model.*;
```
Comment (areklalo) [inline api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java]: I believe you should unwrap this wildcard.

Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java]: I was going to unwrap wildcards from all classes in this repository in my next pull request

Comment (areklalo) [reply api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java]: Ok

**Comment by oskarhinc** [comment]
@pmuchowski I've updated the PR, could you take a look at those changes as well?

File: api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java [hunk 2]
```diff
@@ -136,25 +139,15 @@ public MedicationRequest updateFHIRMedicationRequest(MedicationRequest medicatio
         FHIRMedicationRequestUtil.copyObsAttributes(incomingDrugOrder, generatedDrugOrder, errors);
         if (generatedDrugOrder != null) { //medication request update
             if (!errors.isEmpty()) {
-                StringBuilder errorMessage = new StringBuilder("The request cannot be processed due to the " +
+                String errorMessage = ErrorUtil.generateErrorMessage(errors, "The request cannot be processed due to the " +
```
Comment (kcissewski) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: You can use FHIRConstants.REQUEST_ISSUE_LIST here.

Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: Nica catch, thanks

File: api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java [hunk 3]
```diff
@@ -10,11 +10,14 @@
 import org.openmrs.api.context.Context;
 import org.openmrs.module.fhir.api.util.FHIRMedicationUtil;
 import org.openmrs.module.fhir.api.util.FHIRUtils;
+import org.openmrs.module.fhir.api.util.StrategyUtil;
 import org.springframework.stereotype.Component;
 
 import java.util.ArrayList;
 import java.util.List;
 
+import static org.openmrs.module.fhir.api.util.FHIRUtils.extractUuid;
```
Comment (kcissewski) [inline api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java]: Is it used here?

File: api/src/main/java/org/openmrs/module/fhir/api/util/ErrorUtil.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package org.openmrs.module.fhir.api.util;
+
+import java.util.List;
+
+public class ErrorUtil {
```
Comment (kcissewski) [inline api/src/main/java/org/openmrs/module/fhir/api/util/ErrorUtil.java]: This class can be final and should have private constructor.

File: api/src/main/java/org/openmrs/module/fhir/api/util/StrategyUtil.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package org.openmrs.module.fhir.api.util;
+
+import org.hl7.fhir.dstu3.model.DomainResource;
+import org.hl7.fhir.dstu3.model.IdType;
+
+import java.util.List;
+
+public class StrategyUtil {
```
Comment (kcissewski) [inline api/src/main/java/org/openmrs/module/fhir/api/util/StrategyUtil.java]: This class can be final and should have private constructor.

File: api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java (hunk 3 above)
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java]: No, I can remove it, but as I mentioned to @areklalo I'm going to automatically optimize import in my next PR

File: api/src/main/java/org/openmrs/module/fhir/api/util/ErrorUtil.java (hunk 4 above)
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/util/ErrorUtil.java]: Ok

File: api/src/main/java/org/openmrs/module/fhir/api/util/StrategyUtil.java (hunk 5 above)
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/util/StrategyUtil.java]: Ok

File: api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java (hunk 3 above)
Comment (kcissewski) [reply api/src/main/java/org/openmrs/module/fhir/api/strategies/medication/MedicationStrategy.java]: Sure. I think you can do it later then :)

File: api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java [hunk 6]
```diff
@@ -1,11 +1,230 @@
...
+
+		// Set `ServiceCategory` as EncounterType
+		return diagnosticReport.getCategory().getCoding();
+	}
+
+	protected Patient getPatientFromReport(Encounter omrsDiagnosticReport, Reference subjectReference) {
+		Patient omrsPatient;
+		if (!subjectReference.isEmpty()) {
+			// Get Id of the Patient
+			Identifier patientIdentifier = subjectReference.getIdentifier();
```
[the remark this answers is not in the document]
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/fhir/api/diagnosticreport/handler/AbstractHandler.java]: Ok, thanks, I'll take a look at it on Monday. If you'll merge this PR before I'll look at it in the next one.

**Pull request merged**